### PR TITLE
Automatically created groups with Field to group by as entrytype (#4539)

### PR DESCRIPTION
**Pull request by shubhamatlani**
Potential solution for [#4539](https://github.com/JabRef/jabref/issues/4539):
- Made "entrytype" as a valid keyword for automatically created groups.
- Added check in <code>BibEntry</code> to see if 'searchField' matches TYPE_HEADER - "entrytype"
- Made change in <code>WordKeywordGroup</code> to collect entries for the entrytype groups
- Screenshots reflecting the change attached -

1) All entries -
all_entries_ss1

2) Automatically create groups. Field to group by is "entrytype" -
create_group_ss2

3) Result -
group_view_ss3

----

- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Manually tested changed features in running JabRef
- [ ] Screenshots added in PR description (for bigger UI changes)
- [ ] Ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)

### CODE REVIEW

File: src/main/java/org/jabref/model/groups/WordKeywordGroup.java [hunk 1]
```diff
@@ -113,6 +118,14 @@ public boolean contains(BibEntry entry) {
 
     private Set<String> getFieldContentAsWords(BibEntry entry) {
         if (onlySplitWordsAtSeparator) {
+            if (BibEntry.TYPE_HEADER.equals(searchField)) {
+                for (String searchWord : searchWords) {
+                    Optional<EntryType> entryType = EntryTypes.getType(entry.getType(), BibDatabaseMode.BIBLATEX);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: I think you can move the EntryType.get... Before the loop

Comment (Siedlerchr) [reply src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: Then you could implement the for loop as stream with filter and in the end you can use map / and collectors.toSet

Comment (shubhamatlani) [reply src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: Thanks. I have a question - While fetching entryType, how do I decide which BibDatabaseMode to pass? BIBTEX or BIBLATEX? In both the committed files I have used BIBLATEX. Is that correct?

File: src/main/java/org/jabref/model/entry/BibEntry.java [hunk 2]
```diff
@@ -812,6 +812,15 @@ private void invalidateFieldCache(String fieldName) {
...
             latexFreeFields.put(name, citeKey.get());
             return citeKey;
+        } else if (TYPE_HEADER.equals(name)) {
+            Optional<EntryType> entryType = EntryTypes.getType(getType(), BibDatabaseMode.BIBLATEX);
+            if (entryType.isPresent()) {
+                String entryName = entryType.get().getName();
+                latexFreeFields.put(name, entryName);
+                return Optional.of(entryName);
+            } else {
+                return Optional.empty();
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/entry/BibEntry.java]: In the case when the entry is unknown, you can nonetheless return the value of `getType` as in:
https://github.com/JabRef/jabref/blob/26cd572d4e0587ddf763edc3f9be4caafb1844e6/src/main/java/org/jabref/model/entry/BibEntry.java#L137-L142

File: src/main/java/org/jabref/model/groups/WordKeywordGroup.java [hunk 3]
```diff
@@ -113,6 +118,12 @@ public boolean contains(BibEntry entry) {
 
     private Set<String> getFieldContentAsWords(BibEntry entry) {
         if (onlySplitWordsAtSeparator) {
+            if (BibEntry.TYPE_HEADER.equals(searchField)) {
+                Optional<EntryType> entryType = EntryTypes.getType(entry.getType(), BibDatabaseMode.BIBLATEX);
+                if (entryType.isPresent()) {
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: As above, `entryType = EntryTypes.getType(entry.getType(), BibDatabaseMode.BIBLATEX).orElse(entry.getType())` is slightly better.

File: src/main/java/org/jabref/model/groups/WordKeywordGroup.java [hunk 4]
```diff
@@ -113,6 +118,12 @@ public boolean contains(BibEntry entry) {
...
+                    return searchWords.stream().filter(sw -> entryType.get().getName().equals(sw)).collect(Collectors.toSet());
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: What's the reason for filtering the `searchWords` list? I think you can safely return the value of `entryType` (as a 1-item list).

Comment (shubhamatlani) [reply src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: <code>searchWords</code> is a 1-item set. Filter method lets me compare the content of this set with the entry type. Example, searchWord can be 'Book' but the entry type can be 'Article' and hence, this entry won't fall under Book subgroup. Other approach that I could think of -

<code>if (entryType.isPresent() && searchWords.iterator().next().equals(entryType.get().getName())) {</code>
	<code>return searchWords;</code>
<code>}</code>

Please suggest.

Comment (Siedlerchr) [reply src/main/java/org/jabref/model/groups/WordKeywordGroup.java]: Your solution sounds reasoable, I would let it as is.

**Comment by shubhamatlani** [comment]
Hi, is there anything else that needs to be done here before this PR can be merged with master ?

**Comment by Siedlerchr** [comment]
if @tobiasdiez  is fine with the change, we can merge!

**Pull request merged**